bus-mapping/src/bytecode.rs

```diff
@@ -0,0 +1,165 @@
+    pub fn write_op(&mut self, op: OpcodeId) -> &mut Self {
+        self.write_op_internal(op.as_u8())
+    }
+
+    fn write_op_internal(&mut self, op: u8) -> &mut Self {
+        self.num_opcodes += 1;
+        self.write(op)
+    }
+
+    /// Write byte
+    pub fn write(&mut self, byte: u8) -> &mut Self {
+        self.code.push(byte);
+        self
+    }
+
+    /// Push
+    pub fn push(&mut self, n: usize, value: BigUint) -> &mut Self {
+        assert!((1..=32).contains(&n), "invalid push");
+
+        // Write the op code
+        self.write_op_internal(OpcodeId::PUSH1.as_u8() + ((n - 1) as u8));
+
+        let mut bytes = value.to_bytes_le();
+        // Pad to 32 bytes
+        bytes.resize(32, 0u8);
+        // Write the bytes MSB to LSB
+        for i in 0..n {
+            self.write(bytes[n - 1 - i]);
+        }
+        // Check if the full value could be pushed
+        for byte in bytes.iter().skip(n) {
+            assert!(*byte == 0u8, "value too big for PUSH{}: {}", n, value);
+        }
+        self
+    }
+
+    /// Add marker
+    pub fn add_marker(&mut self, marker: String) -> &mut Self {
+        self.insert_marker(&marker, self.num_opcodes);
+        self
+    }
+
+    /// Insert marker
+    pub fn insert_marker(&mut self, marker: &str, pos: usize) {
+        assert!(
+            !self.markers.contains_key(marker),
+            "marker already used: {}",
+            marker
+        );
+        self.markers.insert(marker.to_string(), pos);
+    }
+
+    /// Get the position of a marker
+    pub fn get_pos(&self, marker: &str) -> usize {
+        *self
+            .markers
+            .get(&marker.to_string())
+            .unwrap_or_else(|| panic!("marker '{}' not found", marker))
+    }
+
+    /// Setup state
+    pub fn setup_state(&mut self) {
+        self.append(&mut crate::bytecode! {
+            PUSH1(0x80u64)
+            PUSH1(0x40u64)
+            MSTORE
+        });
+    }
+
+    /// Call a contract
+    #[allow(clippy::too_many_arguments)]
+    pub fn call(
+        &mut self,
+        gas: BigUint,
+        address: BigUint,
+        value: BigUint,
+        mem_in: BigUint,
+        mem_in_size: BigUint,
+        mem_out: BigUint,
+        mem_out_size: BigUint,
+    ) {
+        self.append(&mut crate::bytecode! {
+            PUSH32(mem_out_size)
+            PUSH32(mem_out)
+            PUSH32(mem_in_size)
+            PUSH32(mem_in)
+            PUSH32(value)
+            PUSH32(address)
+            PUSH32(gas)
+            CALL
+        });
+    }
+}
+
+/// EVM code macro
+#[macro_export]
+macro_rules! bytecode {
+    ($($args:tt)*) => {{
+        let mut code = crate::bytecode::Bytecode::default();
+        $crate::bytecode_internal!(code, $($args)*);
+        code
+    }};
+}
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! bytecode_internal {
+    // Nothing left to do
+    ($code:ident, ) => {};
+    // PUSHX op codes
+    ($code:ident, $x:ident ($v:expr) $($rest:tt)*) => {{
```

Comment:
The `PUSH1(0x1)` vs `PUSH1 0x1` is because the latter is harder to parse without additional delimiters per opcode while having `0x1` as an `expr`. So for example this is easy with additional `;`s to mark the different opcodes:

```
PUSH 0x1;
PUSH 0x2;
```
but then we need a lot of `;` on each opcode line... Unless you have a solution for that?

The `#[start]` vs `.start` no real reason, can be changed if you want. :)